base/special/exp2.jl

```diff
@@ -0,0 +1,446 @@
+exp2_overflow(::Type{Float32}) = 128f0
+exp2_underflow(::Type{Float64}) = -1075.0
+exp2_underflow(::Type{Float32}) = -150f0
+exp2_small(::Type{Float64}) = 2.0^-54
+exp2_small(::Type{Float32}) = 2f0^-24
+@inline exp2_word(x::Float64) = unsafe_trunc(UInt32, (reinterpret(UInt64, x))) # low word
+@inline exp2_word(x::Float32) = reinterpret(UInt32, x) # word
+
+@inline exp2_float_from_words(highword, lowword) = reinterpret(Float64, (UInt64(highword)<<32)|lowword)
+@inline exp2_float_from_words(highword, ::Void ) = reinterpret(Float32, highword)
+
+exp2_minus_large(::Type{Float64}) = 2.0^-1000
+exp2_minus_large(::Type{Float32}) = 2.0^-100
+
+exp2_redux(T::Type{Float64}) = 3*2.0^51 / exp2_tablesize(T)
+exp2_redux(T::Type{Float32}) = 3*2f0^22 / exp2_tablesize(T)
+
+function exp2(x::T) where T<:Union{Float32, Float64}
+    absx = abs(x)
+
+    # Filter out exceptional cases
+    isinf(x) && return ifelse(sign(x) == one(T), x, zero(T))
+    if absx >= exp2_overflow(T)
+        x >= exp2_overflow(T) && return T(Inf) # overflow  without signal
+        x <= exp2_underflow(T) && return zero(T) # underflow without signal
+    elseif absx < exp2_small(T)
+        return one(T) + x
+    end
+
+    # Reduce x, computing z, i0, and k.
+    t = x + exp2_redux(T)
+    i0 = exp2_word(t)
+    i0 += exp2_tablesize(T)>>1
+    k = (i0 >> exp2_tablebits(T)) << 20
+    i0 &= exp2_tablesize(T)-UInt32(1)
+    if T == Float64
+        i0 = i0 << 1
+    end
+    i0 += UInt32(1) # reference index is 1 not 0
+    t -= exp2_redux(T)
+    z = x - t
+
+    # Compute r = exp2(y) and scale by 2^(k>>20)
+    # based on x being Float32 or Float64
+    exp2_kernel(k, i0, z, x)
+end
+
+@inline function exp2_kernel(k, i0, z, x::Float64)
+    # Compute r = exp2(y) = exp2t[i0] * p(z - eps[i]).
+    t = exp2_combined_table[i0]	# exp2t[i0]
+    z -= exp2_combined_table[i0 + 1] # eps[i0]
+    if k >= -(1021 << 20)
+        twopk = exp2_float_from_words(0x3ff00000 + k, 0)
+    else
+        twopkp1000 = exp2_float_from_words(0x3ff00000 + k + (1000 << 20), 0)
+    end
+    r = t + t * z * @horner(z, 0.6931471805599453,
+                               0.2402265069591,
+                               0.0555041086648214,
+                               0.009618129842126066,
+                               0.0013333559164630223)
+
+    # Scale by 2^(k>>20).
+    if k >= -(1021 << 20)
+        if k == 1024 << 20
+            return r * 2.0 * 2.0^1023
+        end
+        return r * twopk
+    else
+        return r * twopkp1000 * exp2_minus_large(T)
+    end
+end
+
+@inline function exp2_kernel(k, i0, z, x::Float32)::Float32
+   twopk = exp2_float_from_words(0x3ff00000 + k, 0)
+
+   #Compute r = exp2(y) = exp2ft[i0] * p(z)
+   @inbounds tv = exp2f_table[i0]
+   u = tv * z
+   tv = tv + u * @horner(z, 0.6931472f0, 0.2402265f0) +
+        u * (z * z) * @horner(z, 0.05550541f0, 0.009618355f0)
+
+   # Scale by 2^(k>>20)
+   return tv * twopk
+end
+
+exp2(x::Real) = exp2(float(x))
```

Comment:
Git, like many other tools, doesn't like too much missing newlines at the end of a file.

Reply:
In what way? I don't think we have any guidelines regarding trailing newlines for this repo.

Reply:
Apparently it is a TODO :smile: 
https://github.com/JuliaLang/julia/blob/53f05292ba7a6f5399ac727bcea1d5c11cdf488e/contrib/check-whitespace.sh#L27

Reply:
Ah right. My old editor did this for me